ECIP-1056 (Agharta): revise draft w/r/t EIP-1702 EVM Account Versioning

meowsbits opened the pull request:
The previous iteration of this document downplayed the introduction
of making the EVM use versioned contract accounts.

This is a complex and significant change and demands adequate attention in the
document and discussion.

This change (including EIP1702 in this fork's "Meta-ECIP") was originally made here: https://github.com/ethereumclassic/ECIPs/commit/4d227a254e69cd1def7765c0d5da91a6606050f1

meowsbits committed:
ECIP-1056 (Agharta): revise draft w/r/t EIP-1702 EVM Account Versioning

The previous iteration of this document downplayed the introduction
of making the EVM use versioned contract accounts.

This is a significant change.

soc1c commented on ECIPs/ecip-1056.md:
```diff
@@ -68,13 +78,21 @@ increases its functionality and should be considered a feature upgrade rather th
+- Geth Classic: no support
+- Parity Ethereum: no support (?)
+- Multi Geth: no support
+- Mantis: no support (?)
```
Mantis does not support Ethereum Classic anymore, so no need to consider it here

meowsbits replied:
Done. https://github.com/etclabscore/ECIPs/pull/35/commits/e3cd5f330525da0984c1da955f3b83e5dee13b75

soc1c commented on ECIPs/ecip-1056.md:
```diff
@@ -10,18 +10,25 @@ created: 2019-03-25
+
 This document proposes the following blocks at which to implement these changes in the Classic networks:
 
 - `1_500_000` on Kotti Classic PoA-testnet (early November 2019)
```
can you also update that?

Potential Agharta timeline

- Morden Testnet (_deprecated_), Dec 11, 2019, 12pm UTC, block `5_348_435` (Gethc, Parity)
- Mordor Testnet, Jan 15, 2020, 12pm UTC, block `617_716` (Gethc, Parity, Mgeth)
- Kotti Testnet, Feb 05, 2020, 12pm UTC, block `2_028_202` (Parity, Mgeth)
- Classic Mainnet, Apr 15, 2020, 12pm UTC, _no block number required yet_

meowsbits replied:
yep, thanks, done, and will close #36  https://github.com/etclabscore/ECIPs/pull/35/commits/6f9cc228031a7347bdd8a991acfb8960218d3370

meowsbits committed:
ECIP-1056: update tentative chain rollout schedules

meowsbits committed:
ECIP-1056: remove Mantis client considerations

The client no longer support ETC via
https://github.com/etclabscore/ECIPs/pull/35/files#r337563231

lialan commented:
Like the authors stated in the proposal, this change is fully backward compatible and is a necessity for future improvement.

My concern here is the storage it takes to support account versioning. A 256-bit scalar is used which will take quite a lot of disk spaces.

Consider using higher bits of the `Nonce` field which is most likely sitting there being all zeros. By doing this we will have almost minimal change to the client. Need inputs from original proposer @sorpaas

sorpaas commented:
@lialan That's not true. RLP encoding is compact, meaning all zero bits are dropped. This means currently it only takes 1-bit for non-legacy versions, and it's simply impossible to use any nonce bits.

sorpaas commented:
I have been suggesting to apply account versioning together with versionless EVM, which has the benefits of making version 1 forward-compatible and thus we will only need 2 versions (compared with one version per six month if only account versioning is applied). You can see my devcon talk for more details on this. If it's proven there're enough interests in ETC community on this, I'd be happy to write more about this.

* ECIP: https://github.com/ethereumclassic/ECIPs/pull/141
* Devcon slides: https://git.that.world/talks/20191011-compatibility.git/plain/presentation.pdf

Also I want to note that I won't be championing any of the specifications for ETC hard forks. It's purely up to ETC community. If you think we shouldn't use account versioning at all -- you're right.

soc1c commented:
Let's discuss this in the call tomorrow.

soc1c commented on ECIPs/ecip-1056.md:
```diff
@@ -10,41 +10,52 @@ created: 2019-03-25
 ### Specification
 
-Enable the following three hard fork features via a new account version `0x1`, as specified in [EIP-1702](https://eips.ethereum.org/EIPS/eip-1702) "Specification" section (commonly known as EIP-1702 variant I):
+Enable the following four hard fork features: 
```
```suggestion
Enable the following four hard fork features:
```

soc1c commented on ECIPs/ecip-1056.md:
```diff
@@ -68,12 +79,18 @@ increases its functionality and should be considered a feature upgrade rather th
-The following clients with Ethereum Classic support implement the _Constaninople_ and _Petersburg_ features currently:
+The following clients with Ethereum Classic support __implement the _Constaninople_ and _Petersburg_ opcodes__ currently:
 
 - Geth Classic: no support
```
```suggestion
 - Geth Classic: all features as per [etclabscore/go-ethereum#40](https://github.com/etclabscore/go-ethereum/issues/40)
```

soc1c commented on ECIPs/ecip-1056.md:
```diff
@@ -68,12 +79,18 @@ increases its functionality and should be considered a feature upgrade rather th
+The following clients with Ethereum Classic support __implement the EIP1702-derived Generalized Account Versioning Scheme__:
+
+- Geth Classic: no support
+- Parity Ethereum: no support (?)
```
```suggestion
- Parity Ethereum: full support as per [paritytech/parity-ethereum#10771](https://github.com/paritytech/parity-ethereum/pull/10771)
```

meowsbits committed:
Remove unnecessary words

Co-Authored-By: Talha Cross <47772477+soc1c@users.noreply.github.com>

lialan commented:
@sorpaas oh right I did not consider the RLP encoding. In that case it looks good.

BelfordZ commented:
Im not sure its been mentioned, but this would make anyone looking to develop a new EVM would need to implement all previous versions as well.

That being said, we must be prudent to ensure that changes are very well documented, and the implementation of each version can be replicated by following this documentation with relative ease. 

We should hold off on including this ECIP until the following are very well defined:

1. What will be the canonical model for describing the EVM
2. In what format will changes to this model be proposed
3. where will the model which specifies each version be located?


Further, it would not be wise to make changes to the EVM until this is sorted out. So to be clear, I don't think we should put a date on this HF until the above is sorted out.

meowsbits committed:
(@soc1c) Update ECIPs/ecip-1056.md

https://github.com/etclabscore/ECIPs/pull/35#discussion_r337897915

Co-Authored-By: Talha Cross <47772477+soc1c@users.noreply.github.com>

meowsbits committed:
(@soc1c) Update ECIPs/ecip-1056.md

https://github.com/etclabscore/ECIPs/pull/35#discussion_r337899323

Co-Authored-By: Talha Cross <47772477+soc1c@users.noreply.github.com>

soc1c commented:
This is obsolete now https://github.com/ethereumclassic/ECIPs/pull/155
